ui/perfherder/CompareSelectorView.jsx

```diff
@@ -0,0 +1,245 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import { react2angular } from 'react2angular';
+import { Container, Col, Row, Button } from 'reactstrap';
+
+import perf from '../js/perf';
+import {
+  getApiUrl,
+  repoEndpoint,
+  getProjectUrl,
+  createQueryParams,
+  resultsetEndpoint,
+} from '../helpers/url';
+import { getData } from '../helpers/http';
+import ErrorMessages from '../shared/ErrorMessages';
+import {
+  compareDefaultTimeRange,
+  prettyErrorMessages,
+  errorMessageClass,
+} from '../helpers/constants';
+import ErrorBoundary from '../shared/ErrorBoundary';
+
+import SelectorCard from './SelectorCard';
+
+// TODO remove $stateParams and $state after switching to react router
+export default class CompareSelectorView extends React.Component {
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      projects: [],
+      failureStatus: null,
+      originalProject: 'mozilla-central',
+      newProject: 'try',
+      originalRevision: '',
+      newRevision: '',
+      errorMessages: [],
+      disableButton: true,
+    };
+    this.submitData = this.submitData.bind(this);
+    this.validateQueryParams = this.validateQueryParams.bind(this);
+    this.validateProject = this.validateProject.bind(this);
+    this.validateRevision = this.validateRevision.bind(this);
+  }
+
+  async componentDidMount() {
+    const { data, failureStatus } = await getData(getApiUrl(repoEndpoint));
+    this.setState({ projects: data, failureStatus });
+    this.validateQueryParams();
+  }
+
+  validateQueryParams() {
+    const {
+      originalProject,
+      newProject,
+      originalRevision,
+      newRevision,
+    } = this.props.$stateParams;
+
+    if (originalProject) {
+      this.validateProject('originalProject', originalProject);
+    }
+
+    if (newProject) {
+      this.validateProject('newProject', newProject);
+    }
```

Comment:
I feel like there *has* to be a better/easier way of validating this data and setting state than these functions, though I'm not sure what it is offhand. This is definitely better than what came before, but it's still overly long and verbose for my liking.

Reply:
Thanks for the review. Yeah, I see your point. I'd be hesitant to add an input/form specific library for one use case (unless @camd thinks it'd be useful elsewhere in the app) but something like a [general validation library](https://validatejs.org/#validate) could kill two birds (form input and query param validation) with one stone. This also looks like it could get a bit verbose but might be worth a shot. Regardless, I was thinking that validation for the compare tab might be best utilized in a higher order component so it can be reused for the different views (compareChooser, compare, comparesubtest). Alternatively, I could move some of these functions into the helper file but they're not that generic (and the graphs and alerts tabs utilize different query parameters in their urls).

Reply:
Thanks for the feedback - I'll look into it more on Monday.

Reply:
@wlach I decided to leave the error handling and validation as-is. I spent some time trying to make the validation/error handling in `CompareSelectorView` more concise but alternative solutions seemed to make the code either unreadable or looked over-engineered. If you have suggestions, please let me know - otherwise I'd rather merge it and move on. As for the `validate.js` library, this would probably be best done later in a new pr so as to change other parts of the app (such as IFV).